java/code/src/com/redhat/rhn/frontend/xmlrpc/system/SystemHandler.java

```diff
@@ -4569,6 +4614,57 @@ public int schedulePackageRemoveByNevra(User loggedInUser, final Integer sid, Li
                 ActionFactory.TYPE_PACKAGES_REMOVE, allowModules)[0].intValue();
     }
 
+    /**
+     * Schedule package lock for a system.
+     *
+     * @param loggedInUser The current user
+     * @param sid ID of the server
+     * @param pkgIdsToLock List of package IDs to lock (as Integers)
+     * @param pkgIdsToUnlock List of package IDs to lock (as Integers)
+     * @param earliestOccurrence Earliest occurrence of the package removal
+     * @return package action id
+     *
+     * @xmlrpc.doc Schedule package installation for a system.
+     * @xmlrpc.param #param("string", "sessionKey")
+     * @xmlrpc.param #param("int", "serverId")
+     * @xmlrpc.param #array_single("int", "pkgIdsToLock")
+     * @xmlrpc.param #array_single("int", "pkgIdsToUnlock")
+     * @xmlrpc.param dateTime.iso8601 earliestOccurrence
+     * @xmlrpc.returntype #param_desc("return_int_success")
+     */
+    public Long schedulePackageLockChange(User loggedInUser, Integer sid,
+                                          List<Integer> pkgIdsToLock,
+                                          List<Integer> pkgIdsToUnlock, Date earliestOccurrence) {
+
+        Server server = SystemManager.lookupByIdAndUser(sid.longValue(), loggedInUser);
+        DataResult<PackageListItem> lockedPackagesResult =
+                PackageManager.systemLockedPackages(server.getId(), null);
+
+        Set<Package> pkgsAlreadyLocked = lockedPackagesResult.stream().map(
+                pkg -> PackageManager.lookupByIdAndUser(pkg.getPackageId(), loggedInUser)).collect(Collectors.toSet());
+
+        Set<Package> pkgsToLock = pkgIdsToLock.stream().map(
+                pkg -> PackageManager.lookupByIdAndUser(pkg.longValue(), loggedInUser)).
+                filter(Objects::nonNull).peek(pkg -> pkg.setLockPending(Boolean.TRUE)).collect(Collectors.toSet());
+
+        Set<Package> pkgsToUnlock = pkgIdsToUnlock.stream().map(
+                        x -> PackageManager.lookupByIdAndUser(x.longValue(), loggedInUser)).
+                filter(Objects::isNull).collect(Collectors.toSet());
+
+        PackageManager.setPendingStatusOnLockedPackages(pkgsToUnlock, PackageManager.PKG_PENDING_UNLOCK);
+
+        PackageManager.lockPackages(server.getId(), pkgsToLock);
+
+        pkgsToLock.addAll(pkgsAlreadyLocked);
```

Comment:
I think you need to reduce `pkgsAlreadyLocked` by `pkgToUnlock`.

`pkgToLock = pkgsAlreadyLocked - pkgsToUnlock + pkgToLock`

or did I miss something?

Reply:
I guess you are right, I'm still testing it. I will put PR in ready when all the changes are finished and tested (currently I put this PR in draft)

Reply:
Actually it seems to be right as it is: `pkgsAlreadyLocked - pkgsToUnlock + pkgToLock` add locks correctly, but packages to be unlock will remain in a `unlocking..` pending status.

Reply:
Check what happens when the action returns. At that point the status should be updated.
SaltUtils.java handlePackageLockData()